core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/AkkaClusterContainerResourceManager.scala

```diff
@@ -0,0 +1,376 @@
+  }
+  def scheduledStopCount = localReservations.values.count {
+    case p: Reservation => p.size.toMB < 0
+    case _              => false
+  }
+  private var clusterResourcesAvailable
+    : Boolean = false //track to log whenever there is a switch from cluster resources being available to not being available
+
+  def canLaunch(memory: ByteSize, poolMemory: Long, poolConfig: ContainerPoolConfig): Boolean = {
+
+    val localRes = localReservations.values.map(_.size) //active local reservations
+    val remoteRes = remoteReservations.values.toList.flatten.map(_.size) //remote/stale reservations
+
+    val allRes = localRes ++ remoteRes
+    //make sure there is at least one node with unreserved mem > memory
+    val canLaunch = clusterHasPotentialMemoryCapacity(memory.toMB, allRes) //consider all reservations blocking till they are removed during NodeStatsUpdate
+    //log only when changing value
+    if (canLaunch != clusterResourcesAvailable) {
+      if (canLaunch) {
+        logging.info(this, s"cluster can launch action with ${memory.toMB}MB reserved:${reservedSize}")
+      } else {
+        logging.warn(this, s"cluster cannot launch action with ${memory.toMB}MB reserved:${reservedSize}")
+      }
+    }
+    clusterResourcesAvailable = canLaunch
+    canLaunch
+  }
+
+  /** Return true to indicate there is expectation that there is "room" to launch a task with these memory/cpu/ports specs */
+  def clusterHasPotentialMemoryCapacity(memory: Double, reserve: Iterable[ByteSize]): Boolean = {
+    //copy AgentStats, then deduct pending tasks
+    var availableResources = clusterActionHostStats.toList.sortBy(_._2.mem).toMap //sort by mem to match lowest value
+    val inNeedReserved = ListBuffer.empty ++ reserve
+
+    var unmatched = 0
+
+    inNeedReserved.foreach { p =>
+      //for each pending find an available offer that fits
+      availableResources.find(_._2.mem > p.toMB) match {
+        case Some(o) =>
+          availableResources = availableResources + (o._1 -> o._2.copy(mem = o._2.mem - p.toMB))
+          inNeedReserved -= p
+        case None => unmatched += 1
+      }
+    }
+    unmatched == 0 && availableResources.exists(_._2.mem > memory)
+  }
+
+  /** reservation adjustments */
+  def addReservation(ref: ActorRef, size: ByteSize): Unit = {
+    localReservations = localReservations + (ref -> Reservation(size))
+  }
+  def releaseReservation(ref: ActorRef): Unit = {
+    localReservations = localReservations - ref
+  }
+  def allowMoreStarts(config: ContainerPoolConfig): Boolean =
+    localReservations
+      .count({ case (_, state) => state.size.toMB > 0 }) < config.clusterManagedResourceMaxStarts //only positive reservations affect ability to start
+
+  class ContainerPoolClusterData(instanceId: InvokerInstanceId, pool: ActorRef) extends Actor {
+    implicit val cluster = Cluster(system)
+    implicit val ec = context.dispatcher
+    val mediator = DistributedPubSub(system).mediator
```

Comment:
Mediator is an impl detail of akka DistributedPubSub; there is a comment below menitoning: ```mediator ! Put(self) //allow point to point messaging based on the actor name: use Send(/user/<myname>) to send messages to me in the cluster
    ```

Reply:
right. my comment was not on what a _mediator_ does, but on its specific _purpose_ for the logic in this PR. Since I had to read bellow to understand that, I though to suggest that the code documents to the reader that only `ReleaseFree` is passed.